### DISCUSSION TITLE
Is it possible to extract raw sql without the "event" middlewares?

### DISCUSSION THREAD

**Discussion by is-it-ayush** [prisma/prisma#17998] [Q&A]
Hi, So I stumbled upon this use-case where I need to get the raw sql query. I know this is possible via events which I can catch in a prisma middleware. However, I need to catch this inside another middleware from a different library. This is my current solution and it makes the use of middlewares. However I'm afraid the event data might get lost since this is a local declaration & if the middleware `foo()` is fired again before the last event could be dealt with. The whole thing is going to break.
```ts
async function foo(ctx) {
     let queryEvent: Prisma.QueryEvent | undefined;
     prisma.$on("query", () (e: Prisma.QueryEvent) => {
           queryEvent = e;
     }
     const result  = await bar();  // <--- fire the actual query...
    
      // at this point we will have our event information in `queryEvent` variable.

      const raw =  queryEvent.query;
      
      // yay! now i have both result & the raw query...
      
}

async function bar(ctx) {
    const query = await prisma.examplemodel.create({..some ctx data});
    return query
}
```
> Note: both foo & bar have access to the same ctx.

So I have two questions,
- Since this code is going to run in a serverless environment. I don't think it'll re-declare the `queryEvent` before I could save it & do something else with the infromation since every `foo() with bar()` is fired in it's own instance (on each request). Am I correct or wrong?
   - Also what is going to happen to local declaration in a non-serverless environment?
- Is it possible to extract the same amount of infromation emitted from `query` event but without the middlewares. (I really don't want to spend time dealing with query engine source to also return the query result (in js object) along with the event, which i don't think i'm even capable of..)

Any help is greatly appreciated. 🙂

**Comment by ludralph** [comment]
Hi @is-it-ayush 👋 

>Since this code is going to run in a serverless environment. I don't think it'll re-declare the queryEvent before I could save it & do something else with the infromation since every foo() with bar() is fired in it's own instance (on each request). Am I correct or wrong?

Yes, you are correct.  In a serverless environment, each function invocation runs in its own isolated container, so you don't need to worry about the queryEvent variable being overwritten by another invocation of foo(). Each function invocation will have its own copy of the variable, and they won't interfere with each other. 

>Also what is going to happen to local declaration in a non-serverless environment?

In a non-serverless environment, you also don't need to worry about the variable being overwritten, since each function call runs in its own call stack and has its own copies of the local variables. However, if you have multiple concurrent requests running in the same process, then you could have multiple instances of foo() running at the same time, which could cause race conditions if they try to access the same queryEvent variable. If you want to avoid this, you could use a closure to encapsulate the queryEvent variable within the foo() function, or use some kind of synchronization mechanism to ensure that only one instance of foo() accesses the queryEvent variable at a time.

For the second question, can you try out this query logging [client extension](https://github.com/prisma/prisma-client-extensions/tree/main/query-logging)

**Reply by is-it-ayush** [reply]
Hi @ludralph, Thank you so much for your response. Could you expand further on "multiple concurrent requests"? I understand how it might lead to race conditions on same process but the former is a bit confusing! I'll crertainly checkout client extensions. Also are you aware of any date until it becomes stable for production?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
